# Music Playlist Database

This project utilizes Udacity's Chinook playlist database. It offers solutions to three straightforward questions, showcasing my SQL skills for data mining. To understand how the tables are interconnected, refer to the database's ERD.

# What are the three most preferred genres?


In [1]:
import sqlite3

connection = sqlite3.connect("chinook.db")
cursor = connection.cursor()

query = """
SELECT
  music_type,
  SUM(avg_total_sales) AS sum_avg_total_sales_by_genre,
  CASE
    WHEN SUM(avg_total_sales) >= 20 THEN 'Top'
    WHEN SUM(avg_total_sales) > 10 AND SUM(avg_total_sales) <= 20 THEN 'Medium'
    ELSE 'Low'
  END AS level_sales
FROM (
  SELECT DISTINCT
    Artist.Name AS artist_name,
    Genre.Name AS music_type,
    COUNT(Album.AlbumId) AS total_album,
    AVG(InvoiceLine.UnitPrice * InvoiceLine.Quantity) AS avg_total_sales
  FROM Artist
  JOIN Album ON Artist.ArtistId = Album.ArtistId
  JOIN Track ON Album.AlbumId = Track.AlbumId
  JOIN Genre ON Track.GenreId = Genre.GenreId
  JOIN InvoiceLine ON Track.TrackId = InvoiceLine.TrackId
  JOIN Invoice ON InvoiceLine.InvoiceId = Invoice.InvoiceId
  GROUP BY Artist.Name
) sub
GROUP BY music_type
ORDER BY sum_avg_total_sales_by_genre DESC;
"""

cursor.execute(query)
results = cursor.fetchall()

for row in results:
    print(row)

connection.close()



('Rock', 47.519999999999996, 'Top')
('Classical', 32.67, 'Top')
('Latin', 23.759999999999998, 'Top')
('Alternative & Punk', 12.87, 'Medium')
('Metal', 9.9, 'Low')
('Jazz', 6.93, 'Low')
('TV Shows', 5.97, 'Low')
('Drama', 3.98, 'Low')
('Blues', 3.96, 'Low')
('World', 2.9699999999999998, 'Low')
('R&B/Soul', 2.9699999999999998, 'Low')
('Alternative', 2.9699999999999998, 'Low')
('Sci Fi & Fantasy', 1.99, 'Low')
('Soundtrack', 1.98, 'Low')
('Reggae', 1.98, 'Low')
('Hip Hop/Rap', 1.98, 'Low')
('Pop', 0.99, 'Low')
('Electronica/Dance', 0.99, 'Low')
('Easy Listening', 0.99, 'Low')
('Bossa Nova', 0.99, 'Low')
('Rock And Roll', 0.9899999999999999, 'Low')


The SQL query retrieves the three most preferred music genres based on total sales from the *Chinook* database. It calculates the sum of average total sales by genre and categorizes each genre into "Top," "Medium," or "Low" sales levels based on the sales thresholds defined in the query. According to the query's output, the top three preferred genres are:

1. **Rock** with total average sales of 47.52 (categorized as "Top").
2. **Classical** with total average sales of 32.67 (categorized as "Top").
3. **Latin** with total average sales of 23.76 (categorized as "Top").

These genres are identified as the most preferred, given their higher sales figures compared to other genres.
This insight is valuable for sales and marketing strategies, as the company can focus more resources and budget on the most popular genres to drive higher sales.

# Which country was the best buyer in 2009?

In [2]:
import sqlite3
connection = sqlite3.connect("chinook.db")
cursor = connection.cursor()
query = """
SELECT
  Customer.Country,
  SUM(InvoiceLine.UnitPrice * InvoiceLine.Quantity) AS Total_sales
FROM Artist
JOIN Album ON Artist.ArtistId = Album.ArtistId
JOIN Track ON Album.AlbumId = Track.AlbumId
JOIN Genre ON Track.GenreId = Genre.GenreId
JOIN InvoiceLine ON Track.TrackId = InvoiceLine.TrackId
JOIN Invoice ON InvoiceLine.InvoiceId = Invoice.InvoiceId
JOIN Customer ON Invoice.CustomerId = Customer.CustomerId
WHERE Invoice.InvoiceDate BETWEEN '2009-01-01' AND '2009-12-26'
GROUP BY Customer.Country
ORDER BY Total_sales DESC;
"""
cursor.execute(query)
results = cursor.fetchall()
for row in results:
    print(row)
connection.close()


('USA', 103.95)
('Canada', 57.42)
('Germany', 53.46)
('Brazil', 37.62)
('France', 34.65)
('United Kingdom', 25.74)
('Poland', 15.84)
('Chile', 15.84)
('Portugal', 11.879999999999999)
('Australia', 11.879999999999999)
('Norway', 10.89)
('Czech Republic', 10.89)
('India', 9.9)
('Netherlands', 8.91)
('Finland', 8.91)
('Ireland', 6.93)
('Belgium', 6.93)
('Sweden', 5.9399999999999995)
('Denmark', 5.9399999999999995)
('Italy', 1.98)
('Austria', 1.98)
('Spain', 0.99)


Based on the query results, we observed that the United States was the leading buyer in 2009, with a total turnover of 104 USD, far exceeding other countries. This figure is 54.8 percent higher than that of Canada, which had 57 USD in sales, making the USA a dominant market for that year. The data also highlights that Canada and Germany are the next top markets.

Therefore, for organizing a concert focused on music from the 2000s, targeting the USA, Canada, and Germany would be strategic for maximizing marketing impact.

# Which was the most popular album ever sold on the playlist?

In [3]:
import sqlite3

connection = sqlite3.connect("chinook.db")
cursor = connection.cursor()
query = """
SELECT 
  Genre.Name AS Music_Genre, 
  Artist.Name AS Artist_Name, 
  Album.Title AS Album_Title, 
  SUM(InvoiceLine.InvoiceId) AS Total_Buy
FROM 
  Artist
JOIN 
  Album ON Artist.ArtistId = Album.ArtistId
JOIN 
  Track ON Album.AlbumId = Track.AlbumId
JOIN 
  Genre ON Track.GenreId = Genre.GenreId
JOIN 
  InvoiceLine ON Track.TrackId = InvoiceLine.TrackId
JOIN 
  Invoice ON InvoiceLine.InvoiceId = Invoice.InvoiceId
JOIN 
  Customer ON Invoice.CustomerId = Customer.CustomerId
GROUP BY 
  Music_Genre, Artist_Name, Album_Title
ORDER BY 
  Total_Buy DESC
LIMIT 15;
"""
cursor.execute(query)
results = cursor.fetchall()

for row in results:
    print(row)
connection.close()


('Rock', 'The Who', 'My Generation - The Very Best Of The Who', 5285)
('Latin', 'Os Paralamas Do Sucesso', 'Acústico MTV', 4409)
('Alternative & Punk', 'Titãs', 'Acústico', 4318)
('Rock', 'Soundgarden', 'A-Sides', 4219)
('Rock', 'U2', 'Rattle And Hum', 4199)
('Jazz', 'Gene Krupa', "Up An' Atom", 4135)
('Rock', 'O Terço', 'Compositores', 3861)
('Sci Fi & Fantasy', 'Battlestar Galactica (Classic)', 'Battlestar Galactica (Classic), Season 1', 3791)
('Alternative & Punk', 'Smashing Pumpkins', 'Rotten Apples: Greatest Hits', 3734)
('Latin', 'Antônio Carlos Jobim', 'Chill: Brazil (Disc 2)', 3672)
('Rock', 'U2', 'B-Sides 1980-1990', 3631)
('Rock', 'Kiss', 'Greatest Kiss', 3571)
('Latin', 'Gilberto Gil', 'Quanta Gente Veio Ver (Live)', 3497)
('Bossa Nova', 'Toquinho & Vinícius', 'Vinícius De Moraes - Sem Limite', 3478)
('Latin', 'Chico Buarque', 'Minha Historia', 3362)


We looked up "best album ever sold since the beginning" in the Chinook database. My Generation - The Very Best Of The Who, which was released in 1996 by the group The Who, turns out to be the album with the greatest sales. With rock being the most popular genre in the database, the music is of the rock type and has collaborations or other insights.

My Generation - The Very Best Of The Who brought in $5,285 in revenue, which is 83% higher than their second-best Latin album, Acústico MTV (4,409 USD).

From a marketing perspective, this might help to promote the group's other albums.


# Who was the employee with the most buys and their favorite genre?

In [4]:
import sqlite3
connection = sqlite3.connect("chinook.db")
cursor = connection.cursor()
query = """
SELECT DISTINCT
  Employee.FirstName,
  Employee.LastName,
  Genre.Name AS Music_Type,
  SUM(InvoiceLine.UnitPrice * InvoiceLine.Quantity) AS Total_Buy
FROM 
  Artist
JOIN 
  Album ON Artist.ArtistId = Album.ArtistId
JOIN 
  Track ON Album.AlbumId = Track.AlbumId
JOIN 
  Genre ON Track.GenreId = Genre.GenreId
JOIN 
  InvoiceLine ON Track.TrackId = InvoiceLine.TrackId
JOIN 
  Invoice ON InvoiceLine.InvoiceId = Invoice.InvoiceId
JOIN 
  Customer ON Invoice.CustomerId = Customer.CustomerId
JOIN 
  Employee ON Customer.SupportRepId = Employee.EmployeeId
GROUP BY 
  Employee.FirstName,
  Employee.LastName,
  Genre.Name
ORDER BY 
  Total_Buy DESC;
"""

cursor.execute(query)
results = cursor.fetchall()

for row in results:
    print(row)
connection.close()


('Jane', 'Peacock', 'Rock', 300.96)
('Margaret', 'Park', 'Rock', 297.0)
('Steve', 'Johnson', 'Rock', 228.69)
('Jane', 'Peacock', 'Latin', 137.60999999999999)
('Margaret', 'Park', 'Latin', 126.72)
('Steve', 'Johnson', 'Latin', 117.81)
('Margaret', 'Park', 'Metal', 96.03)
('Steve', 'Johnson', 'Alternative & Punk', 88.11)
('Jane', 'Peacock', 'Metal', 85.14)
('Margaret', 'Park', 'Alternative & Punk', 83.16)
('Steve', 'Johnson', 'Metal', 80.19)
('Jane', 'Peacock', 'Alternative & Punk', 70.29)
('Jane', 'Peacock', 'TV Shows', 37.81)
('Steve', 'Johnson', 'TV Shows', 35.82)
('Jane', 'Peacock', 'Jazz', 33.66)
('Margaret', 'Park', 'Jazz', 29.7)
('Steve', 'Johnson', 'Blues', 24.75)
('Margaret', 'Park', 'Drama', 21.89)
('Jane', 'Peacock', 'Sci Fi & Fantasy', 19.9)
('Margaret', 'Park', 'TV Shows', 19.9)
('Steve', 'Johnson', 'Drama', 19.9)
('Jane', 'Peacock', 'Blues', 18.81)
('Jane', 'Peacock', 'Classical', 18.81)
('Jane', 'Peacock', 'R&B/Soul', 17.82)
('Margaret', 'Park', 'Blues', 16.83)
('Jane', 'P

 We inquired whether any employees were also customers, and it turns out that among the 8 employees in the company, 3 from the sales team (indicated by names on the bar charts) purchased different music genres.
The top-selling genre was rock music, with Jane Peacock leading the way, having spent $301 in this category.

If the company were to initiate an internal communication campaign, these employees could be rewarded with something like a free pass to purchase a rock album of their choice.